components/merkledb/src/values.rs

```diff
@@ -261,6 +261,42 @@ impl BinaryValue for [u8; HASH_SIZE] {
     }
 }
 
+/// Concatenates buffer slices, prepends each slice with its size.
+pub fn concat_buffers(buffers: &mut [Vec<u8>]) -> Vec<u8> {
+    use std::convert::TryFrom;
+
+    let mut res_buf = Vec::new();
+    for mut buf in buffers {
+        let size = u32::try_from(buf.len()).unwrap_or_else(|_| panic!("Too large buffer"));
+        res_buf.append(&mut size.to_bytes());
+        res_buf.append(&mut buf);
```

Comment:
After doing some local benchmarks, using `append` appears ~2x slower than using `extend_from_slice`. Please use `extend_from_slice`. Also, it will allow to use `size.to_le_bytes()` above (which, unlike `to_bytes`, doesn't allocate on heap). Pre-allocating with `Vec::with_capacity` leads to ~30% performance boost, so I'd strongly suggest to use it as well.